### DISCUSSION TITLE
Can we format CSF (Core flight system) data locally?

### DISCUSSION THREAD

**Discussion by ZedPython** [nasa/cFS#720] [Q&A]
I have a binary dump of housekeeping telemetry data related to DS application. How can I decode the telemetry and learn how it is encoded/decoded? 

All i know is the following:
DS application version 2.5.1
cFE version 6.5.0

Example of binary telemetry dump:

09ab c001 0045 2d03 a05d 94bc 0164 c8ff

96ef 0100 02e5 efff df07 ceff 0000 001f

efff ffff dfff ffff d000 02b8 433a 5c64

732e 7462 6c00 0000 0000 0000 0000 0000

0000 0000 0000 0000 0000 0000 09ab c002

0045 2d03 a05e 94bc 0263 c9fe 97ee 0100

02ec effe df08 cefe 0000 0020 efff fffe

**Comment by irowebbn** [comment]
The data structures for telemetry are conventionally defined in the `_msg.h` public header of an application, so in this case [`DS/fsw/inc/ds_msg.h`](https://github.com/nasa/DS/blob/f04f6a04da60a3a003200cccb99ff71008f97f1e/fsw/inc/ds_msg.h) (in your old version, this is under the [`DS/fsw/src`](https://github.com/nasa/DS/blob/2.5.1/fsw/src/ds_msg.h) folder, but the struct is effectively the same):
```c
typedef struct
{
    uint8  CmdAcceptedCounter;                 /**< \brief Count of valid commands received */
    uint8  CmdRejectedCounter;                 /**< \brief Count of invalid commands received */
    uint8  DestTblLoadCounter;                 /**< \brief Count of destination file table loads */
    uint8  DestTblErrCounter;                  /**< \brief Count of failed attempts to get table data pointer */
    uint8  FilterTblLoadCounter;               /**< \brief Count of packet filter table loads */
    uint8  FilterTblErrCounter;                /**< \brief Count of failed attempts to get table data pointer */
    uint8  AppEnableState;                     /**< \brief Application enable/disable state */
    uint8  Spare8;                             /**< \brief Structure alignment padding */
    uint16 FileWriteCounter;                   /**< \brief Count of good destination file writes */
    uint16 FileWriteErrCounter;                /**< \brief Count of bad destination file writes */
    uint16 FileUpdateCounter;                  /**< \brief Count of good updates to secondary header */
    uint16 FileUpdateErrCounter;               /**< \brief Count of bad updates to secondary header */
    uint32 DisabledPktCounter;                 /**< \brief Count of packets discarded (DS was disabled) */
    uint32 IgnoredPktCounter;                  /**< \brief Count of packets discarded
                                                *
                                                * Incoming packets will be discarded when:
                                                *  - The File and/or Filter Table has failed to load
                                                *  - A packet (that is not a DS HK or command packet) has been received
                                                *    that is not listed in the Filter Table
                                                */
    uint32 FilteredPktCounter;                 /**< \brief Count of packets discarded (failed filter test) */
    uint32 PassedPktCounter;                   /**< \brief Count of packets that passed filter test */
    char   FilterTblFilename[OS_MAX_PATH_LEN]; /**< \brief Name of filter table file */
} DS_HkTlm_Payload_t;

/**
 * \brief Application housekeeping packet
 */
typedef struct
{
    CFE_MSG_TelemetryHeader_t TelemetryHeader; /**< \brief cFE Software Bus telemetry message header */

    DS_HkTlm_Payload_t Payload;
} DS_HkPacket_t;
```

From [`cfe/modules/core_api/fsw/inc/cfe_msg_api_typedefs.h`](https://github.com/nasa/cFE/blob/ee187426d08e0d4f0edf640d07381b8f676be8d3/modules/core_api/fsw/inc/cfe_config_api_typedefs.h), we can see where the typedef points:
`typedef struct CFE_MSG_TelemetryHeader CFE_MSG_TelemetryHeader_t;`

Then, in  `cfe/modules/msg/option_inc/default_cfe_msg_hdr_pri.h`
```c
typedef struct
{
    CCSDS_PrimaryHeader_t Pri; /**< \brief CCSDS Primary Header */
} CCSDS_SpacePacket_t;

/**
 * \brief cFS generic base message
 *
 * This provides the definition of CFE_MSG_Message_t
 */
union CFE_MSG_Message
{
    CCSDS_SpacePacket_t CCSDS;                             /**< \brief CCSDS Header (Pri or Pri + Ext) */
    uint8               Byte[sizeof(CCSDS_SpacePacket_t)]; /**< \brief Byte level access */
};

struct CFE_MSG_TelemetryHeader
{
    CFE_MSG_Message_t                  Msg;      /**< \brief Base message */
    CFE_MSG_TelemetrySecondaryHeader_t Sec;      /**< \brief Secondary header */
    uint8                              Spare[4]; /**< \brief Pad to avoid compiler padding if payload
                                                             requires 64 bit alignment */
};
```

Note that in your old version of cFE, the payload was not aligned to a 64-bit boundary, so the 4 spare bits are not present in the telemetry.

The definition for the packet structure underlying the CFE Primary header can be found at CCSDS Space Packet Protocol: https://public.ccsds.org/Pubs/133x0b2e1.pdf , figure 4.2
![image](https://github.com/nasa/cFS/assets/9010221/646ecedc-e36c-4e7f-93a5-e6b80813dd57)

The corresponding struct definition is at [`cfe/modules/msg/fsw/inc/ccsds_hdr.h`](https://github.com/nasa/cFE/blob/ee187426d08e0d4f0edf640d07381b8f676be8d3/modules/msg/fsw/inc/ccsds_hdr.h):

```c
typedef struct CCSDS_PrimaryHeader
{

    uint8 StreamId[2]; /**< \brief packet identifier word (stream ID) */
                       /*  bits  shift   ------------ description ---------------- */
                       /* 0x07FF    0  : application ID                            */
                       /* 0x0800   11  : secondary header: 0 = absent, 1 = present */
                       /* 0x1000   12  : packet type:      0 = TLM, 1 = CMD        */
                       /* 0xE000   13  : CCSDS version:    0 = ver 1, 1 = ver 2    */

    uint8 Sequence[2]; /**< \brief packet sequence word */
                       /*  bits  shift   ------------ description ---------------- */
                       /* 0x3FFF    0  : sequence count                            */
                       /* 0xC000   14  : segmentation flags:  3 = complete packet  */

    uint8 Length[2]; /**< \brief packet length word */
                     /*  bits  shift   ------------ description ---------------- */
                     /* 0xFFFF    0  : (total packet length) - 7                 */

} CCSDS_PrimaryHeader_t;
```

We can see the secondary header defined at: [`cfe/modules/msg/option_inc/default_cfe_msg_sechdr.h`](https://github.com/nasa/cFE/blob/ee187426d08e0d4f0edf640d07381b8f676be8d3/modules/msg/option_inc/default_cfe_msg_sechdr.h):
```c
/**
 * \brief cFS telemetry secondary header
 */
typedef struct
{

    uint8 Time[6]; /**< \brief Time, big endian: 4 byte seconds, 2 byte subseconds */

} CFE_MSG_TelemetrySecondaryHeader_t;
```
| Raw Hex | Value | Meaning | 
|-----------|-------|------------|
| 0x09ab  |          | Tlm Packet w/ secondary header. APID=0x1AB. Looks like this spacecraft used a different msgid (0x09AB) than the default for DS (0x08B8) |
| 0xc001 |     | Full packet, sequence count 1 |
| 0x0045 | 69 | Total packet length is 69 + 7 = 76 bytes |
| 0x2d03'a05d | 755,212,381 | Time (seconds) (possibly Dec 11, 2003 21:12:48 UTC if using GPS time)
| 0x94bc | 38,076 | Time (subseconds) |
| 0x01 | 1 | Cmd accepted Counter |
| 0x64 | 100 |  Cmd rejected counter |
| 0xc8 | 200 | Dest Tbl Load Counter |
| 0xff | 255 |  Dest Tbl Err Counter | 
| 0x96 | 150 | Filter Tbl Load Counter |
| 0xef | 239 | Filter Tbl Err Counter |
| 0x01 | 1 | App Enable State: Enabled |
| 0x00 | | Spare |
| 0x02e5 | 741 | File Write Counter |
| 0xefff | 61,439 | File Write Err Counter |
| 0xdf07 | 57,095 | File Update Counter |
| 0xceff | 52,991 | File Update Err Counter |
| 0x0000'001f | 31 | Disabled Pkt Counter |
| 0xefff'ffff | 4,026,531,839 | Ignored Pkt Counter |
| 0xdfff'ffff | 3,758,096,383 | Filtered Pkt Counter |
| 0xd000'02b8 | 3,489,661,624 | Passed Pkt Counter |
| 0x433a'5c64'732e'7462'6c00 | "C:\ds.tbl" | Filter Table Filename |

Your second packet starts at the 0x09AB and has a timestamp one second later, but is cut off in the snippet you shared.

For automated decoding, checkout the ground station tool packaged with cFS:
https://github.com/nasa/cFS-GroundSystem/tree/main
